### ISSUE TITLE
MPI Progress Semantics

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#154]
**Originally by _balaji_ on 2009-04-07 22:39:27 -0500**

---
## Description

This ticket is a pre-cursor for some additions that might be required for MPI's interoperability with other programming models.

MPI 2.1 currently does not define what kind of progress rules the underlying implementation follows. That is, the completion of a non-blocking operation might depend on the application calling another MPI function. Even if some MPI implementation provides implicit progress semantics in some cases, the application has to assume the worst case, i.e., no implicit progress.

For example, consider the following example for MPI + Sockets:

```
  if (rank == 0) {
     MPI_Isend(... data to rank 1 ...);
     recv(... from non-MPI socket connected to rank 1 ...)
  }
  else if (rank == 1) {
     MPI_Recv(... data from rank 0 ...)
     send(... to non-MPI socket connected to rank 0 ...)
  }
```

In this case, whether the application would deadlock or complete depends on the MPI implementation makes implicit progress on the MPI_Isend() call. This is not defined by MPI-2.1 and is a quality of implementation issue. The application writer has no way of getting this information.
## Proposed Solution

The proposed solution is to allow the application to query what kind of progress support the MPI implementation provides:

```
   MPI_Progress_semantics(int requested, int *provided, MPI_Info info);
```

"requested" is the user requested level of progress and "provided" is the MPI implementation provided level of progress. Possible values these can take are MPI_PROGRESS_EXPLICIT (no implicit progress), MPI_PROGRESS_IMPLICIT (implicit progress).

MPI_Info allows the user to hint what cases it uses. For example, it can provide the following pre-defined keys:

-_no-derived-datatypes_*: the application does not use any derived datatypes.

-_no-wildcard-source_*: the application does not use any wildcard source.

-_no-wildcard-tag_*: the application does not use any wildcard tag.

-_error-on-explicit-fallback_*: the MPI implementation can return an error if the MPI implementation does implicit progress by default but has to fallback to explicit mode if something internally fails (e.g., failed registration). -- This might be too complicated; we might consider dropping it.

These hints allow the MPI implementation to give a better worst-case estimate.
## Impact on Applications / Users

Currently, interoperability between MPI and other programming models
is mostly undefined (except threads) because of potential deadlocks in cases where calls from other programming models are interleaved with that of MPI, and the MPI implementation does not provide implicit progress. With this function, the application user can query the MPI implementation for the level of support it provides, and use this information to either make interleaved calls (if implicit progress is supported), or fall back to a phase-based approach (e.g., there is a UPC phase followed by a UPC barrier, then an MPI phase followed by an MPI barrier) if implicit progress is not supported. If the application does not want to do that, it can choose to abort when the required level of support is not provided.
## Impact on MPI Implementations

For most MPI implementations this function is just a matter of telling the user what kind of semantics it provides.
## Alternative Solutions

An alternate solution is that the user could set a timer within the application (which spawns a thread to make progress) or have a separate thread make progress. This works well for cluster kind of platforms, but is restrictive for platforms where threads are not allowed (BG/L or BG/P VN mode) or might waste a core for systems with light-weight kernels that do not support multiple processes/threads per processing element (e.g., BG/P dual and smp mode).

Also, this forces MPI to be initialized as threaded (at least funneled), which is not needed when the system has hardware support for implicit progress, or a common network subsystem that multiple programming models rely on for progress (e.g., BG DCMF or Cray Portals).
## Entry for the Change Log

New function to allow MPI implementations to expose the level of communication progress support that they provide.

**Comment by mpiforumbot** [comment]
**Originally by _rbbrigh_ on 2009-04-08 10:00:15 -0500**

---

I have several issues with this proposal.  I don't understand why it's MPI's responsibility to ensure progress for other communication systems.  That is, is there an equivalent UPC routine for registering an MPI progress function so that MPI makes progress when blocked on UPC calls?  To me, this is clearly the responsibility of the network layer below MPI and UPC (or whatever other hybrid programming system is used).  Presumably the layer below can track the outstanding communication requests for all upper-level protocols and make progress on them.  If there is no common layer below (the path to the network is independent), then this functionality is moot anyway.

I can see the most common use of this function being:
  MPI_Register_external_function( MPI_Make_progress, NULL, MPI_OFTEN );

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2009-04-08 10:23:24 -0500**

---

I agree with the comment that it's as much UPC's responsibility as MPI's responsibility to make progress. However, if neither party adds a call, these two would be unusable together.

With respect to the network layer making progress -- how would it know MPI or UPC semantics?

Finally, you cannot do a MPI_Make_progress in the external function as the notes state that it cannot be a blocking or MPI function (otherwise we might have problems with threads nesting).

**Comment by mpiforumbot** [comment]
**Originally by _rbbrigh_ on 2009-04-08 10:55:26 -0500**

---

If the MPI implementation doesn't have independent progress, then you'll need for the other layer to do something similar for MPI progress.  They are unusable together unless both have the ability to make progress while blocked on the other.  You'll essentially need this for **every** communication system that could potentially need to make progress while blocked in any one of them.

The need for the lower layer to know the semantics of the lower layer is not necessarily required.  It depends on what needs to be done to make progress.  To be general, the lower layer could have a function for registering progress functions from the upper layers.  It shouldn't be MPI that collects all of these, it should be the lowest layer in the network.

**Comment by mpiforumbot** [comment]
**Originally by _rbbrigh_ on 2009-04-08 10:58:13 -0500**

---

Whoops. I meant "The need for the lower layer to know the semantics of the upper layer..." in the first sentence of the second paragraph....

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2009-04-08 11:15:30 -0500**

---

I see your point. However, how will the application know whether the MPI implementation (or UPC implementation) will have such implicit progress? If there is a possibility of a deadlock, I (application) would prefer to know it early.

Note that saying "it's MPI's responsibility" might not be an accurate statement as this is only a quality of implementation issue for MPI (and presumably for UPC as well). Further, from an MPI's perspective, how would it know what underlying network UPC would use? E.g., on the same system, MPI could use IB and UPC could use TCP.

An alternative proposal could be to request the MPI implementation to tell the application what kind of progress semantics it provides (e.g., explicit, implicit). Something along the lines of:

```
   MPI_Progress_info(requested, &provided);
```

This would require other programming models to provide such semantics as well, though. I can initiate some conversation with the UPC folks if this seems better.

I think, though, that Global Arrays will always guarantees implicit progress, so the application might not need to make any checks in that case. Jim: can you confirm this part?

**Comment by mpiforumbot** [comment]
**Originally by _rbbrigh_ on 2009-04-08 11:40:34 -0500**

---

I guess my point is that this is an implementation detail that shouldn't be exposed to the application.  Whether or not the underlying network provides progress is an implementation detail that only the MPI implementor knows.  If the underlying network wants to support multiple upper-level protocols that could potentially interfere with each other, then it has to have a way of mitigating that.  If it can't provide the necessary progress semantics, then it needs to have a way of registering progress for all of the upper-level protocols.  The MPI implementor would register an internal MPI implementation function that allows it to progress all outstanding operations.  The UPC implementor would have to do the same thing.  You really don't want to rely on the application to know or understand that it's mixing multiple network layers that need to kick the others when appropriate.

You're example with MPI/IB and UPC/TCP is what I meant by independent paths through the network.  If there's no common underlying network, then I think this issue is moot.  You can't make it the responsibility of only one layer to take responsibility for all of the others because they all need to know about each other.

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2009-04-08 20:36:19 -0500**

---

I have modified the ticket (significantly) to capture all the discussion we had on the ticket as well as separately (thanks Ron and Darius). Please comment on how the current version of the ticket sounds.

**Comment by mpiforumbot** [comment]
**Originally by _jdinan_ on 2009-04-09 13:20:15 -0500**

---

Replying to balaji:

> I have modified the ticket (significantly) to capture all the discussion we had on the ticket as well as separately (thanks Ron and Darius). Please comment on how the current version of the ticket sounds.

I don't think this will be sufficient for eliminating deadlock with Berkeley UPC's progress semantics.  The problem with BUPC is that one-sided calls can, in effect, be two-sided in their implementation.  Meaning that the remote node must call a upc runtime function or explicitly call bupc_poll().  Even on rDMA networks where get and put are truly one-sided, upc_lock(), for example, may still require cooperation from the remote node.

Because of this, a one-sided upc communication can block the matching MPI call that a remote node is waiting on causing circular wait.  No amount of MPI progress will fix this situation:

```
Process 0:

upc_memget(...From process 1...)  // Blocked here, waiting on UPC progress
MPI_Send(...To process 1...)

Process 1:

MPI_Recv(...From process 0...) // Blocked here, not making UPC progress
```

I agree that this it's not truly MPI's responsibility to guarantee progress for external libraries and programming models.  But it would be a highly useful feature.  Global Arrays and ARMCI, for example, spawn a data server thread on each node to guarantee implicit progress of one-sided operations.  Because of this, there are no restrictions on mixing these programming models with MPI.  However, on some systems spawning a thread to manage communication might not be feasible.

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2009-04-09 13:27:36 -0500**

---

There are two aspects here --- one that is needed from MPI and the other that is needed from UPC. We can do the MPI part. As Ron pointed out, and I agree with, it is incorrect to change the MPI standard to accommodate inefficiencies in UPC. If UPC does not do implicit progress, it should either change its model to always do it (like Global Arrays) or provide a call to the user to tell whether it would or would not (as this proposal does for MPI).

Also, just to clarify, here we are not trying to do anything special to avoid deadlocks. We are just querying the MPI implementation about whether it does implicit progress or not. This might be a required condition to avoid deadlocks, but is certainly not sufficient. UPC will have to do its part too, for MPI + UPC to work together.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-05-05 12:21:09 -0500**

---

I generally agree that it might be a good idea to query the asynch. progress support. However, we then get into the discussion of different "progression levels". What we have right now seems suboptimal. For example "implicit progress only for contiguous data communication" - that's exposing MPI internal things to the user that he should not have to worry about. I understand this from an implementation standpoint, however, the selection seems random. What about vector collectives? I don't have a better proposal off-hand. Even for the "contiguous data progress", I see problems with rendezvous protocols in non-threaded environments. Maybe we should just go with two layers?

This whole issue is even worse, nobody prohibits the MPI library to copy the passed value in a temp buffer and not send it immediately (Open MPI/openib did this thing at some point), thus, all those discussions also apply to blocking communication.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-05-26 12:32:01 -0500**

---

There was discussion about this ticket on the hybrid teleconf today.

One point that occurred to me after the teleconf is that not even BSD sockets gives you this kind of information because, just like MPI, the answer is almost always "it depends."  Many factors come into play, but most importantly, some factors aren't known until you actually try to send.

So you can't necessarily know ahead of time whether progress will be made independently of calling other MPI functions _until the transfer is done_ -- just like you can't know if write(2) will block until you actually try to call it.

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2009-05-26 19:51:45 -0500**

---

I agree that the MPI implementation can only respond with the least common denominator value here. For example, if it has to fall back to explicit progress even in a rare case, it will have to return MPI_PROGRESS_EXPLICIT. However, for an MPI implementation that uses an asynchronous thread or an active messages kind of model (using interrupts, for example), it might still be able to say that it provides implicit progress.

Another point that Jim Dinan brought up after the call today is that this function also gives a hint to the MPI implementation that the application needs implicit progress, in which case the MPI library can start an additional thread, for example, to do background progress.

I think the biggest concern here is whether there are enough cases where many MPI implementations can give the answer as MPI_PROGRESS_IMPLICIT. If in 99.9% cases, the answer is going to be MPI_PROGRESS_EXPLICIT, then this addition will probably not be worth it.

I'm also going to change the ticket to take an MPI_Info argument to let the user tell the implementation what all features the application cares about (e.g., derived datatypes, wildcard source/tag).

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2009-06-08 17:52:44 -0500**

---

Updated the ticket to add the MPI_Info argument and some additional changes as discussed in the last hybrid PM working group.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2009-06-23 11:34:11 -0500**

---

moved it per wg telecon to not ready (on hold).

**Comment by mpiforumbot** [comment]
**Originally by _rlgraham_ on 2010-09-18 04:09:26 -0500**

---

Pavan, what do you want to do with this ticket ?

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2010-09-19 08:31:23 -0500**

---

The Forum is clearly against this. Not pursuing this further.
